### PR TITLE
`while_let_on_iterator` Improvements

### PR DESCRIPTION
**Pull request by Jarcho**
fixes: #6491
fixes: #6231
fixes: #5844
fixes: #1924
fixes: #1033

The check for whether a field can be borrowed should probably be moved to utils at some point, but it would require some cleanup work and knowing what parts can actually be shared.

changelog: Suggest `&mut iter` when the iterator is used after the loop.
changelog: Suggest `&mut iter` when the iterator is a field in a struct.
changelog: Don't lint when the iterator is a field in a struct, and the struct is used in the loop.
changelog: Lint when the loop is nested in another loop, but suggest `&mut iter` unless the iterator is from a local declared inside the loop.

### CODE REVIEW

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (xFrednet) [inline clippy_lints/src/loops/while_let_on_iterator.rs:12]: Personally I would prefer a name like `match_expr` instead of `scrutinee_expr`. _Scrutinee_ is the correct therm, but I had to look it up while `match_expr` would ensure that everybody understands it right away.

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (xFrednet) [inline clippy_lints/src/loops/while_let_on_iterator.rs:32]: This as well as the previous match statement could be written using an `if_chain!` with `then {} else { return }` blocks. This can improve readability since it checks only one condition per line and reduce nesting. Another advantage of this is also that new conditions can easily be added. Here is an example rewrite:
```rust
use if_chain::if_chain;

let iter_expr = if_chain!{
    if let ExprKind::MethodCall(name, _, [iter_expr], _) = scrutinee_expr.kind;
    if name.ident.name == sym::next;
    if is_trait_method(cx, scrutinee_expr, sym::Iterator);
    if let Some(iter_expr) = try_parse_iter_expr(cx, iter_expr);
    then {
        iter_expr
    } else {
        return;
    }
};
```

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (xFrednet) [inline clippy_lints/src/loops/while_let_on_iterator.rs:43]: This `if` results in a return if `some_pat` is `None`. You could move this test to the `some_pat` extraction in, line 13 if you use `if_chain!`. This also ensures that you don't have to deal with an `Option<>` here and later on :)

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (xFrednet) [inline clippy_lints/src/loops/while_let_on_iterator.rs:58]: This can be simplified since we know from line 43 that `some_pat` is `Some`.

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (xFrednet) [inline clippy_lints/src/loops/while_let_on_iterator.rs:139]: Could you clarify the first sentence a bit? I'm unsure what you mean with it.

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (xFrednet) [inline clippy_lints/src/loops/while_let_on_iterator.rs:169]:
```suggestion
        // If the path matches, this is either an exact match or the expression is a parent of the field.
```

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (xFrednet) [inline clippy_lints/src/loops/while_let_on_iterator.rs:180]: With functions like this that return a tuple with simple base types it's usually good to explain what they mean. I had to read the code and think a bit to figure out that it returns true if a field or path has been skipped. :)

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (xFrednet) [inline clippy_lints/src/loops/while_let_on_iterator.rs:233]: You could extract the nested visitors into their own functions to reduce the line count. That would also reduce the scope a reader has to comprehend at once. Keeping this as it is would also be fine as it's still very understandable IMO.

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (xFrednet) [inline clippy_lints/src/loops/while_let_on_iterator.rs:260]: A simple `=` is enough since this code can only be reached if `self.used_iter` is `false` (Lines: 249 - 251).
```suggestion
                    self.used_iter = is_res_used(self.cx, self.iter_expr.path, id);
```

File: clippy_utils/src/visitors.rs
Comment (xFrednet) [inline clippy_utils/src/visitors.rs:192]:
```suggestion
/// Checks if the given resolved path is used inside the given body.
```

File: tests/ui/while_let_on_iterator.rs
Comment (xFrednet) [inline tests/ui/while_let_on_iterator.rs:275]: lintcheck also found examples where the loop was writing to a different field of the struct. In this case it would have an expression like this: `self.0.1 = <some expr>`. Could you also add a test loop for this? :upside_down_face:

**Review by xFrednet — CHANGES_REQUESTED** [review]
Hey, this is my first bigger review in Clippy. I therefor also annotated some NITs which are probably fine. The work is really impressive, and you put a lot of effort into this! The implementation looks good beside the suggestions. I'm not completely done with the tests yet, but they are on the to-do list for the weekend.

A thing that is not directly connected to your changes: Could you also extend the documentation of the lint itself a bit? The [current documentation](https://rust-lang.github.io/rust-clippy/master/index.html#while_let_on_iterator) is very basic. Even extending it by the way the example _should_ be written as would be a great improvement IMO.

@Manishearth Could you look over these suggestions? :upside_down_face: 

---

I've also tested the lint using lintcheck this added 3 new lint triggers which seem valid and the suggestions (I've tested) also seem to work fine :+1: 

<details>
<summary>The lint triggers with the linted code</summary>

1. `proc-macro2-1.0.24/src/parse.rs:444:5`
    ```rust
    while let Some((i, ch)) = chars.next() {
        match ch {
            '"' => {
                n = i;
                break;
            }
            '#' => {}
            _ => return Err(LexError),
        }
    }
    ```
2. `serde-1.0.118/src/private/de.rs:2830`
    ```rust
    while let Some(item) = self.iter.next() {
        // Items in the vector are nulled out when used by a struct.
        if let Some((ref key, ref content)) = *item {
            self.pending_content = Some(content);
            return seed.deserialize(ContentRefDeserializer::new(key)).map(Some);
        }
    }
    ```
3. `serde-1.0.118/src/private/de.rs:2932`
   ```rust
   while let Some(item) = self.iter.next() {
        if let Some((ref key, ref content)) = *item {
            // Do not take(), instead borrow this entry. The internally tagged
            // enum does its own buffering so we can't tell whether this entry
            // is going to be consumed. Borrowing here leaves the entry
            // available for later flattened fields.
            self.pending = Some(content);
            return seed.deserialize(ContentRefDeserializer::new(key)).map(Some);
        }
    }
   ```

</details>

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (Jarcho) [reply clippy_lints/src/loops/while_let_on_iterator.rs:12]: `match_expr` could be confused for the actual `match` expression. Probably not in this case, but it is ambiguous to name it like that.

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (Jarcho) [reply clippy_lints/src/loops/while_let_on_iterator.rs:32]: rustfmt skips `if_chain` calls, so this does come with a downside. I guess that objection isn't really applicable in this case.

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (xFrednet) [reply clippy_lints/src/loops/while_let_on_iterator.rs:12]: Fair point :upside_down_face:

**Comment by Manishearth** [comment]
@xFrednet your review looks great so far! Nice catches!

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (Jarcho) [reply clippy_lints/src/loops/while_let_on_iterator.rs:43]: It does, but it shouldn't.

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (Jarcho) [reply clippy_lints/src/loops/while_let_on_iterator.rs:139]: Added an example there.

File: tests/ui/while_let_on_iterator.rs
Comment (Jarcho) [reply tests/ui/while_let_on_iterator.rs:275]: A test like that already exists (line 288-292). I guess it could prevent a future regression if someone makes a weird change though.

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (Jarcho) [reply clippy_lints/src/loops/while_let_on_iterator.rs:227]: I'm on the fence on this one. It's not really too long (100 lines isn't really a good limit) and it makes it clear the visitors are only used in the function. On the other hand the start of the function is a little ways down.

Comment (xFrednet) [reply clippy_lints/src/loops/while_let_on_iterator.rs:227]: I believe that 100 lines are usually a good limit since it's not too common to nest structs and entire implementation in functions. So I'm fine with leaving it as it is :)

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (xFrednet) [reply clippy_lints/src/loops/while_let_on_iterator.rs]: Ah, you've combined the two `match` statements into a single `if_chain!`. I thought more about implementing them as two chains. This looks good at a first glance, so what ever works ^^. I'll try to review all changes and finish the review this week, however it'll probably be pushed into the weekend :upside_down_face:

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (Jarcho) [reply clippy_lints/src/loops/while_let_on_iterator.rs:227]: Then I'll mark this as resolved.

File: clippy_lints/src/loops/while_let_on_iterator.rs
Comment (Jarcho) [reply clippy_lints/src/loops/while_let_on_iterator.rs]: There isn't really any reason to have two chains one after another. I did return from the chain rather than put everything into the `then` block so rustfmt will touch it.

Comment (Jarcho) [reply clippy_lints/src/loops/while_let_on_iterator.rs]: If you have an unambiguous name other than scrutinee I can switch to it. Personally I think we should switch to using it everywhere so everyone can learn it.

Comment (xFrednet) [reply clippy_lints/src/loops/while_let_on_iterator.rs]: I think it would be good if we use the correct therms for the syntax, the only problem is that it can increase the learning curve for newcomers to the project. That was also the main reason for the suggestion your arguments are compelling, and I was happy to learn a new therm. So, you can resolve this thread as well, at least from my side :)

Comment (Jarcho) [reply clippy_lints/src/loops/while_let_on_iterator.rs]: Sounds good.

**Comment by xFrednet** [comment]
Sorry for taking so long for the second review, I've been busy due to an assignment that I have to hand in on the 4th. This PR is high on the TODO list after that handin :upside_down_face:

**Review by xFrednet — CHANGES_REQUESTED** [review]
Thank you for adding your changes in a second commit that helped with the review a lot and sorry, for taking so long for the second review :sweat_smile:.

I only have one more small NIT where it would be nice if you could fix that even if it's not part of your changes. The rest looks really solid and can be merged after a rebase IMO. In don't know if @Manishearth wants to do a final review.

---

The small NIT is in [`test/ui/while_let_on_iterator.rs#L151-L159`](https://github.com/rust-lang/rust-clippy/blob/b208a809ec08dd5b4200ed826d05d758f129a061/tests/ui/while_let_on_iterator.rs#L151-L159):
```rust
    // This should not cause an ICE and suggest:
    //
    // for _ in values.iter() {}
    //
    use std::collections::HashSet;
    let mut values = HashSet::new();
    values.insert(1);

    while let Some(..) = values.iter().next() {}
```
This code doesn't trigger the lint as the iterator is always newly generated. Could you remove the *suggestion* part of the comment above?

**Comment by Manishearth** [comment]
Your review looks great, I'll let you r+ when ready!

@bors delegate=xFrednet

**Comment by xFrednet** [comment]
The final update looks good. Thank you!
@bors r+

**Pull request merged**
